Get Package info

AlexAxthelm opened the pull request:
This PR is a checkpoint PR, and should be considered as part of a larger work in progress, mostly containing the get_package_info and related functions.

I'm opening this as a targeted PR, so that this bit can be evaluated in isolation, since the tests for this section can get kind of gnarly.

Requests to reviewers:
* Ensure that the information emitted in the package info is what we want (not missing anything)
* Help check that the tests are actually testing what I think they are

Note that a lot of the "weirdness" in the tests (making use of a lot of `expect_match` vs `expect_identical`, or the heavy use of for example) is related to either having the tests run on the CI machines, or run on Windows, or because I don't actually want to break `.liPaths()` on my local machine 😆

AlexAxthelm commented:
@cjyetman @jdhoffa I still have a few tests to add for `get_package_info`, but `get_individual_package_info` is ready for review.

cjyetman commented:
one nit is that I'd rather it return a list without the package name built into it, e.g.
``` r
devtools::load_all()
#> ℹ Loading pacta.workflow.utils
pkg_name <- "pacta.data.preparation"

# would rather do this
pacta.workflow.utils:::get_individual_package_info(pkg_name)$remotepkgref
#> NULL

# than this
pacta.workflow.utils:::get_individual_package_info(pkg_name)[[pkg_name]]$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation#305"
```

other than that it seems pretty robust for `pak::pak()` installs...
``` r
devtools::load_all()
#> ℹ Loading pacta.workflow.utils
pkg_name <- "pacta.data.preparation"

remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0("rmi-pacta/", pkg_name), ask = FALSE))
(pkg_info <- pacta.workflow.utils:::get_individual_package_info(pkg_name))
#> $pacta.data.preparation
#> $pacta.data.preparation$package
#> [1] "pacta.data.preparation"
#> 
#> $pacta.data.preparation$version
#> [1] "0.1.0.9003"
#> 
#> $pacta.data.preparation$library
#> [1] "/Users/cjrmi/Library/R/arm64/4.3/library"
#> 
#> $pacta.data.preparation$library_index
#> [1] 1
#> 
#> $pacta.data.preparation$repository
#> [1] NA
#> 
#> $pacta.data.preparation$platform
#> [1] "*"
#> 
#> $pacta.data.preparation$built
#> [1] "R 4.3.2; ; 2024-03-21 18:51:27 UTC; unix"
#> 
#> $pacta.data.preparation$remotetype
#> [1] "github"
#> 
#> $pacta.data.preparation$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation"
#> 
#> $pacta.data.preparation$remoteref
#> [1] "HEAD"
#> 
#> $pacta.data.preparation$remotesha
#> [1] "8b2d052b7d3b988c31f3ca2248e92091539be788"
remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0(pkg_info[[pkg_name]]$remotepkgref), ask = FALSE))
pacta.workflow.utils:::get_individual_package_info(pkg_name)[[pkg_name]]$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation"


sha <- "f8a2418c"
remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0("rmi-pacta/", pkg_name, "@", sha), ask = FALSE))
(pkg_info <- pacta.workflow.utils:::get_individual_package_info(pkg_name))
#> $pacta.data.preparation
#> $pacta.data.preparation$package
#> [1] "pacta.data.preparation"
#> 
#> $pacta.data.preparation$version
#> [1] "0.1.0.9003"
#> 
#> $pacta.data.preparation$library
#> [1] "/Users/cjrmi/Library/R/arm64/4.3/library"
#> 
#> $pacta.data.preparation$library_index
#> [1] 1
#> 
#> $pacta.data.preparation$repository
#> [1] NA
#> 
#> $pacta.data.preparation$platform
#> [1] "*"
#> 
#> $pacta.data.preparation$built
#> [1] "R 4.3.2; ; 2024-03-21 18:58:20 UTC; unix"
#> 
#> $pacta.data.preparation$remotetype
#> [1] "github"
#> 
#> $pacta.data.preparation$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation@f8a2418c"
#> 
#> $pacta.data.preparation$remoteref
#> [1] "f8a2418c"
#> 
#> $pacta.data.preparation$remotesha
#> [1] "f8a2418c86456536e5f708ced46d6fb608afecd1"
remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0(pkg_info[[pkg_name]]$remotepkgref), ask = FALSE))
pacta.workflow.utils:::get_individual_package_info(pkg_name)[[pkg_name]]$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation@f8a2418c"


branch <- "add-prepare_abcd_flags_"
remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0("rmi-pacta/", pkg_name, "@", branch), ask = FALSE))
(pkg_info <- pacta.workflow.utils:::get_individual_package_info(pkg_name))
#> $pacta.data.preparation
#> $pacta.data.preparation$package
#> [1] "pacta.data.preparation"
#> 
#> $pacta.data.preparation$version
#> [1] "0.1.0.9003"
#> 
#> $pacta.data.preparation$library
#> [1] "/Users/cjrmi/Library/R/arm64/4.3/library"
#> 
#> $pacta.data.preparation$library_index
#> [1] 1
#> 
#> $pacta.data.preparation$repository
#> [1] NA
#> 
#> $pacta.data.preparation$platform
#> [1] "*"
#> 
#> $pacta.data.preparation$built
#> [1] "R 4.3.2; ; 2024-03-21 18:59:53 UTC; unix"
#> 
#> $pacta.data.preparation$remotetype
#> [1] "github"
#> 
#> $pacta.data.preparation$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation@add-prepare_abcd_flags_"
#> 
#> $pacta.data.preparation$remoteref
#> [1] "add-prepare_abcd_flags_"
#> 
#> $pacta.data.preparation$remotesha
#> [1] "6e5a14455d4bae707bd7727ea469731302433f8b"
remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0(pkg_info[[pkg_name]]$remotepkgref), ask = FALSE))
pacta.workflow.utils:::get_individual_package_info(pkg_name)[[pkg_name]]$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation@add-prepare_abcd_flags_"


tag <- "v0.1.0"
remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0("rmi-pacta/", pkg_name, "@", tag), ask = FALSE))
(pkg_info <- pacta.workflow.utils:::get_individual_package_info(pkg_name))
#> $pacta.data.preparation
#> $pacta.data.preparation$package
#> [1] "pacta.data.preparation"
#> 
#> $pacta.data.preparation$version
#> [1] "0.1.0"
#> 
#> $pacta.data.preparation$library
#> [1] "/Users/cjrmi/Library/R/arm64/4.3/library"
#> 
#> $pacta.data.preparation$library_index
#> [1] 1
#> 
#> $pacta.data.preparation$repository
#> [1] NA
#> 
#> $pacta.data.preparation$platform
#> [1] "*"
#> 
#> $pacta.data.preparation$built
#> [1] "R 4.3.2; ; 2024-03-21 19:01:10 UTC; unix"
#> 
#> $pacta.data.preparation$remotetype
#> [1] "github"
#> 
#> $pacta.data.preparation$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation@v0.1.0"
#> 
#> $pacta.data.preparation$remoteref
#> [1] "v0.1.0"
#> 
#> $pacta.data.preparation$remotesha
#> [1] "15d7f2da8939287e9c7c588b15cbe45a3ea7ba9a"
remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0(pkg_info[[pkg_name]]$remotepkgref), ask = FALSE))
pacta.workflow.utils:::get_individual_package_info(pkg_name)[[pkg_name]]$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation@v0.1.0"


pr <- "305"
remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0("rmi-pacta/", pkg_name, "#", pr), ask = FALSE))
(pkg_info <- pacta.workflow.utils:::get_individual_package_info(pkg_name))
#> $pacta.data.preparation
#> $pacta.data.preparation$package
#> [1] "pacta.data.preparation"
#> 
#> $pacta.data.preparation$version
#> [1] "0.1.0.9003"
#> 
#> $pacta.data.preparation$library
#> [1] "/Users/cjrmi/Library/R/arm64/4.3/library"
#> 
#> $pacta.data.preparation$library_index
#> [1] 1
#> 
#> $pacta.data.preparation$repository
#> [1] NA
#> 
#> $pacta.data.preparation$platform
#> [1] "*"
#> 
#> $pacta.data.preparation$built
#> [1] "R 4.3.2; ; 2024-03-21 19:01:42 UTC; unix"
#> 
#> $pacta.data.preparation$remotetype
#> [1] "github"
#> 
#> $pacta.data.preparation$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation#305"
#> 
#> $pacta.data.preparation$remoteref
#> [1] NA
#> 
#> $pacta.data.preparation$remotesha
#> [1] "7f55c1daf738906eb82c40f37e3ced5d1a764ef4"
remove.packages(pkg_name)
#> Removing package from '/Users/cjrmi/Library/R/arm64/4.3/library'
#> (as 'lib' is unspecified)
suppressMessages(pak::pak(paste0(pkg_info[[pkg_name]]$remotepkgref), ask = FALSE))
pacta.workflow.utils:::get_individual_package_info(pkg_name)[[pkg_name]]$remotepkgref
#> [1] "rmi-pacta/pacta.data.preparation#305"
```

cjyetman commented:
would be curious how robust it is with local instals, e.g. git status etc.

AlexAxthelm commented:
@cjyetman 

> without the package name built into it

Sure. I think that's not a problem.

> would be curious how robust it is with local instals, e.g. git status etc.

Thanks for reminding me of this. I'm thinking that I might add one more key into the list before this is all said and done. For local installs (`remotetype == "local"`) it would be good to get the git status of the repo? Or maybe identifying anything loaded with `devtools::load_all()`? I haven't written the git parts yet, but the `gert` package is looking promising there. It seems to be a solid alternative to `git2r`, but easier to use (and `git2r` has an `R > 4.0` requirement that `gert` doesn't.)

> other than that it seems pretty robust for `pak::pak()` installs...

In this big block, would you like any of these as tests? I've implemented catching refs in one of the tests, but I didn't permute on the different ref types that `pak` can grab.

It would be pretty easy to add more tests (or different repos), but I'm not inclined to effectively re-test `pak`.

https://github.com/RMI-PACTA/pacta.workflow.utils/blob/d0228c255c7ae50fecfcec66ca4411ec0911c3ca/tests/testthat/test-get_individual_package_info.R#L241-L259

I made this function in the tests, and I'm finding it actually quite handy to be able to play with "disposable" local installs, or to inject a development version temporarily higher on the search path than my "actual" (usually from `main`) install. 

https://github.com/RMI-PACTA/pacta.workflow.utils/blob/d0228c255c7ae50fecfcec66ca4411ec0911c3ca/tests/testthat/test-get_individual_package_info.R#L154-L162

cjyetman commented:
I think git status is useful if we're still installing from local. We may be getting away from that though, now that almost everything is in a public repo. But there may be cases where we still want to test local changes we've made to a repo, and in those situations I think it's good to test local git status so we don't get confused.

I don't think any of those things in my test code need to be tested. I just wanted to check if all those variations that `pak` allows come out properly, and can then be reused to reinstall the same thing. I was basically curious if `remotepkgref` always comes out as a valid reference that can be used to install.

Checking for being loaded with `devtools::load_all()` is kind of interesting, but we should probably stop doing that anywhere in "production", and only be using that during development.

Definitely approve of using `withr`. My example was quick and dirty, but using `withr` is much safer and cleaner.

jdhoffa commented:
I think that @cjyetman is the more critical reviewer here, but this all LGTM!

cjyetman commented:
I just reviewed `get_individual_package_info()` as suggested, but I'll try to review the rest of it later today.

AlexAxthelm commented:
@cjyetman, @jdhoffa Made a number of changes since last review, but I think the big thing that's still missing on this block is the git info for local/dev packages, and I haven't written any of that yet, so I'd like to get this merged into the bigger tracking branch.

structural changes since last review:
* `get_package_info` takes an arbitrarily nested list/vector/(anything that can go into vapply) and spits out a nested list.
  * please look at the structures in the tests. there is a little weirdness if you're passing in an unnamed list, but I think the defaults are sensible.
* That's useful since `get_individual_package_info` changed the output structure, and all the useful stuff is now top-level keys
* That function also added a big if/else block dependent on if the package was loaded with `pkgload` (`devtools`), but with this, we'll have a lot more power diagnosing systems that are in local development on restricted/airgapped systems (@jacobvjk this might be relevant for the `multiloanbook` repo)
* Added a few dependencies. mostly suggests.

jdhoffa commented:
NIT: Should we align on/ be more specific about what we mean by `dev_version`?
Currently we are defining it as something different to what is idiosyncratic in R. In typical R packages, the concept of "dev version" is indicated by the large component in SemVer (e.g. `X.Y.Z.9000`) (regardless of if it is loaded using `devtools::load_all()` or installed from GH using e.g. `pak::pak`.

However, here we are defining `dev_version` as something that was loaded into the environment using `devtools::load_all` (see reprex).

Should we rename the flag as `loaded_version` or something similar to distinguish it? 

(I don't mean to start a whole big discussion on what "dev version" _should_ mean, just flagging that it may be confusing to users, or even to us down the road, as to what we are referring to when we use that term)

``` r
devtools::load_all("/Users/jdhoffa/github/pacta.workflow.utils")
#> ℹ Loading pacta.workflow.utils

out <- pacta.workflow.utils:::get_package_info(
  packagelist = c(
    "pacta.workflow.utils",
    "r2dii.data",
    "dplyr"
  )
)
#> Warning in get_individual_package_info(x): Identifying development packages may
#> not be accurate.
#> Warning in get_individual_package_info(x): Multiple installations of package
#> found.

#> Warning in get_individual_package_info(x): Multiple installations of package
#> found.

out$pacta.workflow.utils$version
#> [1] "DEV 0.0.0.9002"
out$pacta.workflow.utils$dev_version
#> [1] TRUE

out$r2dii.data$version
#> [1] "0.4.1.9000"
out$r2dii.data$dev_version
#> [1] FALSE

out$dplyr$version
#> [1] "1.1.4"
out$dplyr$dev_version
#> [1] FALSE
```

<sup>Created on 2024-03-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

AlexAxthelm commented:
@jdhoffa if we renamed it to something like `loaded_with_pkgload` would that be better? It would be more accurate. `local_dev` might also work, but still feels like it's missing something.

I'd stay away from `loaded_version`, since that might be exactly the same version number as what's in the installed-dev (vs the local working dev version). For example, I can install a package from GitHub, change a few files (but not DESCRIPTION), and `load_all` and the version number would be the same (this is the main reason why I think version-bumping on every commit to `main` is useful)

jdhoffa commented:
Yeah that's exactly my point actually! 
I agree that the case it is trying to distinguish between is very useful, just being a bit nitpicky about the name. 

I think `loaded_with_pkgload` sounds reasonable?

AlexAxthelm commented:
@jdhoffa easy change. done. I think I like this clarity better.

jdhoffa reviewed (approved):
That was my only comment, this looks absolutely awesome man seriously!
Thank you 

Leave it to @cjyetman who may have more useful feedback as he engages with the `manifest` outputs more.

cjyetman commented on R/get_environment.R:
```diff
@@ -0,0 +1,30 @@
+get_r_session_info <- function() {
+  return(
+    list(
+      R.version = utils::sessionInfo()[["R.version"]],
```
```r
$R.version$nickname
[1] "Eye Holes"
```
🤣

jdhoffa replied:
LOL

cjyetman commented on tests/testthat/test-get_environment.R:
```diff
@@ -0,0 +1,32 @@
+## save current settings so that we can reset later
+threshold <- logger::log_threshold()
+appender  <- logger::log_appender()
+layout    <- logger::log_layout()
+on.exit({
+  ## reset logger settings
+  logger::log_threshold(threshold)
+  logger::log_layout(layout)
+  logger::log_appender(appender)
+})
+
+logger::log_appender(logger::appender_stdout)
+logger::log_threshold(logger::FATAL)
+logger::log_layout(logger::layout_simple)
```
shame you can't [with_envvar()](https://withr.r-lib.org/reference/with_envvar.html) or  [with_options()](https://withr.r-lib.org/reference/with_options.html) here

AlexAxthelm replied:
Probably can set up a helper, but that's for when I start refactoring tests to also test logging.

I'm imagining something along the lines of 


```
helper_suppress_logs()

test_that("test_1" {expect_foo()})

test_that("test 2", {
  expect_log(
    x <- foo(),
    debug = "foo!"
  )
})

cjyetman commented on tests/testthat/test-get_environment.R:
```diff
@@ -0,0 +1,32 @@
+logger::log_threshold(logger::FATAL)
+logger::log_layout(logger::layout_simple)
+
+test_that("get_single_file_metadata processes CSV tables correctly", {
```
seems like this test name/message is out of sync with reality
